src/mlpack/tests/imputation_test.cpp

```diff
@@ -0,0 +1,323 @@
+ *
+ * Tests for data::Imputer class
+ */
+#include <sstream>
+
+#include <mlpack/core.hpp>
+#include <mlpack/core/data/dataset_mapper.hpp>
+#include <mlpack/core/data/map_policies/increment_policy.hpp>
+#include <mlpack/core/data/map_policies/missing_policy.hpp>
+#include <mlpack/core/data/imputer.hpp>
+#include <mlpack/core/data/imputation_methods/custom_imputation.hpp>
+#include <mlpack/core/data/imputation_methods/listwise_deletion.hpp>
+#include <mlpack/core/data/imputation_methods/mean_imputation.hpp>
+#include <mlpack/core/data/imputation_methods/median_imputation.hpp>
+
+#include <boost/test/unit_test.hpp>
+#include "test_tools.hpp"
+
+using namespace mlpack;
+using namespace mlpack::data;
+using namespace std;
+
+BOOST_AUTO_TEST_SUITE(ImputationTest);
+
+/**
+ * Make sure a CSV is loaded correctly.
+ */
+BOOST_AUTO_TEST_CASE(DatasetMapperImputerTest)
+{
+  fstream f;
+  f.open("test_file.csv", fstream::out);
+  f << "a, 2, 3"  << endl;
+  f << "5, 6, a"  << endl;
+  f << "8, 9, 10" << endl;
+  f.close();
+
+  arma::mat input;
+  arma::mat output;
+
+  std::set<string> mset;
+  mset.insert("a");
+  MissingPolicy policy(mset);
+  DatasetMapper<MissingPolicy> info(policy);
+  BOOST_REQUIRE(data::Load("test_file.csv", input, info) == true);
+
+  // row and column test
+  BOOST_REQUIRE_EQUAL(input.n_rows, 3);
+  BOOST_REQUIRE_EQUAL(input.n_cols, 3);
+
+  // Load check
+  // MissingPolicy should convert strings to nans
+  BOOST_REQUIRE(std::isnan(input(0, 0)) == true);
+  BOOST_REQUIRE_CLOSE(input(0, 1), 5.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(input(0, 2), 8.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(input(1, 0), 2.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(input(1, 1), 6.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(input(1, 2), 9.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(input(2, 0), 3.0, 1e-5);
+  BOOST_REQUIRE(std::isnan(input(2, 1)) == true);
+  BOOST_REQUIRE_CLOSE(input(2, 2), 10.0, 1e-5);
+
+  CustomImputation<double> customStrategy(99); // convert missing vals to 99.
+  Imputer<double,
+          DatasetMapper<MissingPolicy>,
+          CustomImputation<double>> imputer(info, customStrategy);
+  imputer.Impute(input, output, "a", 0); // convert a -> 99 for dimension 0
+
+  // Custom imputation result check
+  BOOST_REQUIRE_CLOSE(output(0, 0), 99.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(output(0, 1), 5.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(output(0, 2), 8.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(output(1, 0), 2.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(output(1, 1), 6.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(output(1, 2), 9.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(output(2, 0), 3.0, 1e-5);
+  BOOST_REQUIRE(std::isnan(output(2, 1)) == true); // remains as NaN
+  BOOST_REQUIRE_CLOSE(output(2, 2), 10.0, 1e-5);
+
+  // Remove the file.
+  remove("test_file.csv");
+}
+
+/**
+ * Make sure a CSV is loaded correctly.
+ */
+BOOST_AUTO_TEST_CASE(CustomImputationTest)
+{
+  arma::mat input("3.0 0.0 2.0 0.0;"
+                  "5.0 6.0 0.0 6.0;"
+                  "9.0 8.0 4.0 8.0;");
+  arma::mat outputT; // assume input is transposed
```

Comment:
This comment is a little bit confuse since the name of the parameter already change to columnMajor, we can use a bool value to tell the readers the output is columnMajor or rowMajor.

ex : 

```
bool columnMajor = true;
CustomImputation<double> imputer(customValue);
imputer.Impute(input, outputT, mappedValue, 0/*dimension*/, columnMajor );
```